core-common/src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java

```diff
@@ -128,21 +126,24 @@ public void releaseProducer(final Producer<byte[], byte[]> producer) {
      *
      * @param producer Producer instance to terminate.
      */
-    public void terminateProducer(final Producer<byte[], byte[]> producer) {
-        LOG.info("Received signal to terminate producer " + producer);
+    public void terminateProducer(final String key, final Producer<byte[], byte[]> producer) {
+        LOG.info("Received signal to terminate producer for key {}: {} ", key, producer);
         rwLock.writeLock().lock();
         try {
-            final int index = activeProducers.indexOf(producer);
-            if (index >= 0) {
-                activeProducers.set(index, null);
+            if (activeProducerByKey.containsKey(key)) {
+                activeProducerByKey.remove(key);
```

Comment:
may use only remove, it returns `null if the map contained no mapping for the key`, otherwise it is double contains

Reply:
and since it is not that often operation maybe we can just remove by value, that would anticipate changing the method signature